### DISCUSSION TITLE
Is it possible to get DCAP quote collateral?

### DISCUSSION THREAD

**Discussion by jasl** [gramineproject/gramine#820] [Q&A]
I need to eager prepare quote collateral for verify the quote on non-sgx environment

**Comment by kailun-qin** [comment]
Yes, it's possible via DCAP qpl public APIs (out of Gramine's scope), e.g., `sgx_ql_get_quote_verification_collateral()`. Please kindly refer to https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf pg.20 for further details, thanks!

**Comment by dimakuv** [comment]
@jasl Are you looking for a way to do it *in Gramine* specifically? As @kailun-qin mentioned, it is not possible in current Gramine.

But if you have a real usage for this feature, we could implement a pseudo-file `/dev/attestation/quote-collateral` that would contain this. Though I'm not sure if this is that useful on its own really...

**Comment by jasl** [comment]
@dimakuv @kailun-qin 
I just make a quick chat with Intel SDK team, they give me a possible solution:

- I need modify QVE to expose `tee_get_fmspc_ca_from_quote`
- I need link QPL and QVE with my Gramine app
- I can use `tee_get_fmspc_ca_from_quote`and `tee_get_verification_endorsement` to get collateral

I haven't try yet, what do you think?

**Reply by kailun-qin** [reply]
IMO, this solution is just a temporal workaround (only for urgent needs) and it doesn't seem to require specific changes in Gramine (but rather in your app and related libs).

However, given your real use case is indeed trying to seek a way to get the collateral within Gramine, would you pls kindly share and elaborate more on your scenario with us? In that way, we can further evaluate the need of implementing a pseudo-file `/dev/attestation/quote-collateral` as @dimakuv suggested, which can eventually facilitate your case.

BTW, I've internally synced, DCAP qvl will probably add a public API for this requirement in the following releases - we can then support it in Gramine. cc @dimakuv

**Reply by dimakuv** [reply]
@jasl Why is this so complicated? Why can't you just use `sgx_ql_get_quote_verification_collateral()`?

**Reply by hyjiang** [reply]
> @jasl Why is this so complicated? Why can't you just use `sgx_ql_get_quote_verification_collateral()`?

Because we need a way to get `fmspc` and `ca` before retrieving verification collaterals, they are index to get collaterals.

**Reply by hyjiang** [reply]
> @dimakuv @kailun-qin I just make a quick chat with Intel SDK team, they give me a possible solution:
> 
> * I need modify QVE to expose `tee_get_fmspc_ca_from_quote`
> * I need link QPL and QVE with my Gramine app
> * I can use `tee_get_fmspc_ca_from_quote`and `tee_get_verification_endorsement` to get collateral
> 
> I haven't try yet, what do you think?

As I said in the sync meeting, this workaround works in theory, and it's you app's logic, which doesn't depend on Gramine.
But note that the workaround is a little tricky, you should only treat it as an backup option. 
In addition, we will add additional API in DCAP QVL to help user retrieving collaterals in later release.

**Reply by dimakuv** [reply]
@hyjiang I see, thanks for the explanations.

Indeed, `sgx_ql_get_quote_verification_collateral()` requires FMSPC and CA inputs, which are currently cannot be extracted by any DCAP QVL API.

**Reply by jasl** [reply]
@dimakuv @kailun-qin 

I'm from https://github.com/Phala-Network/phala-blockchain team, our goal is providing a serverless-like distributed computing network.
We are using SGX to secure computing node which based on Gramine, called [PRuntime](https://github.com/Phala-Network/phala-blockchain/tree/master/standalone/pruntime)
and we're using blockchain as control plane to mange computing nodes' lifecyle (register, start, stop, remove, healthy check, etc.), and tasks distribution.

We're using Remote Attestation to generate quote from computing node, and verify it on blockchain to ensure the node is a valid SGX enclave,
currently we supported EPID, now I'm working on DCAP.
The blockchain is [Substrate-based](https://github.com/paritytech/substrate) which provides a WASM environment so it's an untrusted environment and no external IO, so we have to prepare everything before do the verification.

here's an example that how we verify EPID report https://github.com/Phala-Network/phala-blockchain/blob/master/pallets/phala/src/registry.rs#L405-L483

**Reply by jasl** [reply]
> > @dimakuv @kailun-qin I just make a quick chat with Intel SDK team, they give me a possible solution:
> > 
> > * I need modify QVE to expose `tee_get_fmspc_ca_from_quote`
> > * I need link QPL and QVE with my Gramine app
> > * I can use `tee_get_fmspc_ca_from_quote`and `tee_get_verification_endorsement` to get collateral
> > 
> > I haven't try yet, what do you think?
> 
> As I said in the sync meeting, this workaround works in theory, and it's you app's logic, which doesn't depend on Gramine. But note that the workaround is a little tricky, you should only treat it as an backup option. In addition, we will add additional API in DCAP QVL to help user retrieving collaterals in later release.

Great! thanks, for the timing I'll try workaround first, and will migrate to new official release

BTW, is there ETA for the new release? in September maybe?

**Reply by mkow** [reply]
> BTW, is there ETA for the new release? in September maybe?

Likely in around 2 weeks.

**Reply by dimakuv** [reply]
@mkow is talking about the upcoming release of Gramine -- it is likely in around 2 weeks (end of August).

We don't know about the new release of Intel SGX DCAP libraries though. @hyjiang should answer this question.

**Reply by mkow** [reply]
Oops, right, I was only skimming through the discussion and noticed someone asking about the release, so I quickly replied missing the context. Sorry :)
